### PR TITLE
Convert to ewc-plugin structure. Add forcing generation

### PR DESCRIPTION
**Pull request by BSchilperoort**
This PR overhauls this repository.

Added:

- Wflow.jl forcing generation
  - ERA5
  - CMIP
- eWaterCycle model for Wflow.jl (local only)
- example notebooks.

**Note for reviewers:** Just look at the new repo structure and files https://github.com/eWaterCycle/pywflow/tree/ewatercycle_plugin Basically everything is new.
Too many files have been removed to make the "files changed" tab useful.

<hr>

To do after this pr:

- [ ] Rename repo to "ewatercycle-wflowjl"
- [ ] Add package to PyPI.

### CODE REVIEW

**Comment by sverhoeven** [comment]
> @sverhoeven Wflow.jl now uses the grid types "scalar", "points" and "unstructured". I guess changes will have to be made in ewatercycle to support those in e.g. `model.get_value` and `model.get_value_as_xarray`?

You overwrote those methods in src/ewatercycle_wflowjl/model.py:WflowJl class, but indeed moving those to ewatercycle would be nicer for the next non-rectilinear plugin. Created https://github.com/eWaterCycle/ewatercycle/issues/389 .

File: src/ewatercycle_wflowjl/model.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+        with config_file.open(mode="w") as f:
+            f.write(toml.dumps(self._config))
+
+        return config_file
+
+
+def install_wflow():
+    """Install Wflow.jl with the BMI branch."""
+    jl.seval("import Pkg")
+    jl.Pkg.add(name="Wflow", rev="BMI")
```
Comment (sverhoeven) [inline src/ewatercycle_wflowjl/model.py:207]: If the BMI branch be merged at some point, we should add note somewhere to replace the branch with a tag.

File: src/ewatercycle_wflowjl/model.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+    jl.Pkg.add("BasicModelInterface")
+
+
+def check_wflow_install():
+    """Check if Wflow is installed in the Juliacall environment."""
+    try:
+        jl.seval("using Wflow")
+    except JuliaError(match="not found in current path"):
```
Comment (sverhoeven) [inline src/ewatercycle_wflowjl/model.py]: This is not how python works. Replaced match with if/else.

File: src/ewatercycle_wflowjl/model.py [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+        self._config["output"]["path"] = str(
+            self._cfg_dir / self._config["output"]["path"]
+        )
+        self._config["csv"]["path"] = str(self._cfg_dir / self._config["csv"]["path"])
+        self._config["netcdf"]["path"] = str(
+            self._cfg_dir / self._config["netcdf"]["path"]
+        )
+
+        # Enable the following variables for the BMI:
+        self._config["API"]["components"] = [
```
Comment (sverhoeven) [inline src/ewatercycle_wflowjl/model.py:190]: The config file I have was missing API section. I added created API section if it does not exist.

**Review by sverhoeven — APPROVED** [review]
Works for me.
I was able to install (editable mode) and run the notebooks.

I used parameter set from dcache.
The model was running quit slow so I did not run full period.

File: docs/wflowjl_local.ipynb
[the remark this answers is not in the document]
Comment (Peter9192) [reply docs/wflowjl_local.ipynb:1]: @sverhoeven see email I send you earlier today

File: src/ewatercycle_wflowjl/model.py (hunk 3 above)
Comment (BSchilperoort) [reply src/ewatercycle_wflowjl/model.py:190]: Thanks! I should have caught that,

File: docs/generate_cmip_forcing.ipynb
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply docs/generate_cmip_forcing.ipynb:1]: ESMValTool seems to have large memory issues on saving the cubes. With some adjustment in the Dask configuration of esmvaltool I was able to reduce the occurrence on my machine.

File: src/ewatercycle_wflowjl/model.py (hunk 2 above)
Comment (BSchilperoort) [reply src/ewatercycle_wflowjl/model.py]: Yeah I got confused with pytest's raises() here :man_facepalming: 
```py
with pytest.raises(ValueError, match=r".* 123 .*"):
```

**Comment by Peter9192** [comment]
@BSchilperoort @sverhoeven can this be merged?

**Pull request merged**